contrib/playlib/src/mill/playlib/RouterModule.scala

```diff
@@ -0,0 +1,122 @@
+package mill
+package playlib
+
+import coursier.{Cache, MavenRepository}
+import mill.api.Loose
+import mill.playlib.api.RouteCompilerType
+import mill.scalalib.Lib.resolveDependencies
+import mill.scalalib._
+import mill.scalalib.api._
+
+trait RouterModule extends mill.Module with ScalaModule {
+
+  /**
+    * Defines the version of playframework to be used to compile this projects
+    * routes.
+    */
+  def playVersion: T[String]
+
+  /**
+    * The scala version to use for this project. The version defaults to
+    * the value chosen by play but some version are crossbuilt for multiple
+    * versions of scala.
+    * Override it if the default doesn't match your current usage.
+    */
+  override def scalaVersion: T[String] = T {
+    playVersion() match {
+      case _ => "2.12.8" // default version for play 2.6.x and 2.7.x
+    }
+  }
+
+  override def generatedSources = T{ super.generatedSources() ++ Seq(compileRouter().classes) }
+
+  /**
+    * The [[PathRef]] to the main routes file.
+    *
+    * This is the default path for play projects and it should be fine but you
+    * can override it if needed.
+    */
+  def routesFile: T[PathRef] = T {
```

Comment:
I think, this should be a `T.sources`, but am not 100% sure, because it is only a file and not a directory.

http://www.lihaoyi.com/mill/page/tasks.html#sources

Reply:
It may be necessary, I have to investigate how the play routes compiler handles route file inclusion (https://www.playframework.com/documentation/2.6.x/SBTSubProjects) as it seems that the included file is not compiled at the moment. I'll update the pr when I understand how this works.

Reply:
Without the `.sources` it will at least never recompile after a change, as mill will cache it until next clean (which evicts the cache). The `.sources` forces mill to always run and check for changes.

Reply:
ok I had a look and it seems T.sources is going to be needed and I do have to handle a list of files for it to behave correctly

Reply:
T.sources only seems to accept directories. in my case the directory is shared by multiple files only some of which are consumed by the routes compiler. If I understand correctly this may end up triggering unneeded compilations if some other files change around the.

I tried working around this by : 

* defining a setting `routesDirectory` which is a T.sources
* defining a private setting `routesFiles` which is a T[PathRef] and selects the proper files from the `routesDirectory` (it also makes sure they appear in the proper order so the main routes file is processed last)
* making the compileRouter task depend on the `routesFiles` target

If I understand correctly, any change to any files under `routesDirectory` will trigger a trivial reevaluation of routesFiles. The output of routesFiles will not change unless a route file has changed which will prevent useless compilations and should cost less than fully compiling the router.
I hope I got this right as it would be very inconvenient to have to clean the cache manually to force a router recompile after the

Reply:
It should work as you have outlined. I agree, we should not be forced to implement such workarounds in mill for such a common use case. Maybe @lihaoyi can shed some light how to best tackle single source files?